### PR TITLE
[SPARK-14702][SUBMIT] Expose SparkLauncher's ProcessBuilder for user flexibility

### PR DESCRIPTION
**Pull request by a10y**
## What changes were proposed in this pull request?

Expose the `ProcessBuilder` in `SparkLauncher`, as well as opening up `SparkLauncher#startApplication` to allow accepting arbitrary ProcessBuilders.

For more info see [https://issues.apache.org/jira/browse/SPARK-16511](this JIRA ticket)

Also covers [https://issues.apache.org/jira/browse/SPARK-14702](SPARK-14702)
## How was this patch tested?

Unit tests + spark shell

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/launcher/LauncherBackendSuite.scala [hunk 1]
```diff
@@ -17,15 +17,16 @@
...
 
+
 import java.util.concurrent.TimeUnit
 
 import scala.concurrent.duration._
 import scala.language.postfixOps
 
-import org.scalatest.Matchers
 import org.scalatest.concurrent.Eventually._
+import org.scalatest.Matchers
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/launcher/LauncherBackendSuite.scala]: Please avoid making unnecessary changes to imports. This one, in particular, is wrong and would trigger a style checker error if the style checker didn't have a bug...

File: core/src/test/scala/org/apache/spark/launcher/LauncherBackendSuite.scala [hunk 2]
```diff
@@ -17,15 +17,16 @@
 
 package org.apache.spark.launcher
 
+
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/launcher/LauncherBackendSuite.scala]: don't add this extra line

File: launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java [hunk 3]
```diff
@@ -418,14 +414,26 @@ public SparkAppHandle startApplication(SparkAppHandle.Listener... listeners) thr
...
     }
 
-    String loggerPrefix = getClass().getPackage().getName();
-    String loggerName = String.format("%s.app.%s", loggerPrefix, appName);
-    ProcessBuilder pb = createBuilder().redirectErrorStream(true);
+    ProcessBuilder pb = createProcessBuilder().redirectErrorStream(true);
+    return startApplication(appName, pb, listeners);
+  }
+
+  public SparkAppHandle startApplication(String appName, ProcessBuilder pb,
```
Comment (vanzin) [inline launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java]: Style is:

```
public Type method(
    Type arg1,
    Type arg2,
    ...) {
```

File: launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java [hunk 4]
```diff
@@ -418,14 +414,26 @@ public SparkAppHandle startApplication(SparkAppHandle.Listener... listeners) thr
...
+
     pb.environment().put(LauncherProtocol.ENV_LAUNCHER_PORT,
       String.valueOf(LauncherServer.getServerInstance().getPort()));
     pb.environment().put(LauncherProtocol.ENV_LAUNCHER_SECRET, handle.getSecret());
+
+    String loggerPrefix = getClass().getPackage().getName();
+    String loggerName = String.format("%s.app.%s", loggerPrefix, appName);
     try {
-      handle.setChildProc(pb.start(), loggerName);
+      Process proc = pb.start();
```
Comment (vanzin) [inline launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java]: This change is not necessary.

**Comment by vanzin** [comment]
This PR should really be referencing SPARK-14702, which is the older bug. SPARK-16511 should be closed as a duplicate of it.

File: launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java [hunk 5]
```diff
@@ -418,14 +414,26 @@ public SparkAppHandle startApplication(SparkAppHandle.Listener... listeners) thr
...
+    ChildProcAppHandle handle = LauncherServer.newAppHandle();
+    for (SparkAppHandle.Listener l : listeners) {
+      handle.addListener(l);
+    }
+
     pb.environment().put(LauncherProtocol.ENV_LAUNCHER_PORT,
       String.valueOf(LauncherServer.getServerInstance().getPort()));
     pb.environment().put(LauncherProtocol.ENV_LAUNCHER_SECRET, handle.getSecret());
+
+    String loggerPrefix = getClass().getPackage().getName();
```
Comment (vanzin) [inline launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java]: There's an issue now that the user might call `.redirectOutput` on the `ProcessBuilder` and this code might break. So the log redirection really should only happen if the `ProcessBuilder` is configured to redirect to `ProcessBuilder.Redirect.PIPE`. And you'd need to make sure that `redirectErrorStream()` is set in that case.

An alternative approach would be to only do this log redirection from the existing `startApplication` call that does not take a `ProcessBuilder`. In that case, output of the child process for the new `startApplication` should be handled by the caller, just like for the old `launch()` API.

I really wish you could create new `ProcessBuilder.Redirect` values so that we could create a new `LOG` one and fix this properly, but it doesn't look like that's possible.

**Comment by vanzin** [comment]
ok to test

**Comment by vanzin** [comment]
I tried playing a little bit with what this API would look like, and I'm starting to question my previous idea that exposing the ProcessBuilder is the way to go here...

The above issue with logging redirection is just one source of issues. There are other problems, such as the API becoming a little bit convoluted:

```
SparkLauncher launcher = ...;
ProcessBuilder pb = launcher.createProcessBuilder();
launcher.startApplication(pb);
```

And all the different ways to start the Spark app (3 different methods in SparkLauncher + `ProcessBuilder.start()`). At this point I'm starting to think it might be better to mirror parts of the ProcessBuilder API that are interesting. e.g., have:

```
  SparkLauncher directory(File directory)
  SparkLauncher redirectErrorStream(boolean redirectErrorStream)
  SparkLauncher redirectError(ProcessBuilder.Redirect destination)
  SparkLauncher redirectOutput(ProcessBuilder.Redirect destination)
```

Optionally these (since you can use `Redirect.to(File)`):

```
  SparkLauncher redirectError(File destination)
  SparkLauncher redirectOutput(File destination)
```

And add this one which implements the current logger redirection:

```  
  SparkLauncher redirectToLog(String loggerName)
```

By default logging redirection would be done when using `startApplication`, using the current semantics, unless the user has overridden that by calling one of the new methods (which would also apply to `launch`).

This adds more methods and is a bit more work, but it avoids certains oddities in the API, avoids overloading `startApplication`, and hides ProcessBuilder APIs we don't want to expose (like `command()`). What do you think?

**Comment by andreweduffy** [comment]
That sounds reasonable, I can take a shot at that today, I'll submit it as a second PR and then we can close this one out.
